internal/core/src/index/NgramInvertedIndex.cpp

```diff
@@ -0,0 +1,152 @@
+// Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+#include "index/NgramInvertedIndex.h"
+#include "exec/expression/Expr.h"
+
+namespace milvus::index {
+NgramInvertedIndex::NgramInvertedIndex(const storage::FileManagerContext& ctx,
+                                       const NgramParams& params)
+    : min_gram_(params.min_gram), max_gram_(params.max_gram) {
+    schema_ = ctx.fieldDataMeta.field_schema;
+    field_id_ = ctx.fieldDataMeta.field_id;
+    mem_file_manager_ = std::make_shared<MemFileManager>(ctx);
+    disk_file_manager_ = std::make_shared<DiskFileManager>(ctx);
+
+    if (params.loading_index) {
+        path_ = disk_file_manager_->GetLocalNgramIndexPrefix();
+    } else {
+        path_ = disk_file_manager_->GetLocalTempNgramIndexPrefix();
+        boost::filesystem::create_directories(path_);
+        d_type_ = TantivyDataType::Keyword;
+        std::string field_name =
+            std::to_string(disk_file_manager_->GetFieldDataMeta().field_id);
+        wrapper_ = std::make_shared<TantivyIndexWrapper>(
+            field_name.c_str(), path_.c_str(), min_gram_, max_gram_);
+    }
+}
+
+void
+NgramInvertedIndex::BuildWithFieldData(const std::vector<FieldDataPtr>& datas) {
```

Comment:
is it guaranteed that BuildWithFieldData will be called exactly once? if not, index_build_begin_ may be set multiple times.